db/keyval/etcd/plugin_impl_etcd.go

```diff
@@ -76,42 +85,45 @@ func (plugin *Plugin) Init() (err error) {
 	}
 	// Uses config file to establish connection with the database
 	plugin.connection, err = NewEtcdConnectionWithBytes(*etcdClientCfg, plugin.Log)
-	if err != nil {
-		plugin.Log.Errorf("Err: %v", err)
-		return err
-	}
-	plugin.reconnectResync = etcdCfg.ReconnectResync
-	if etcdCfg.AutoCompact > 0 {
-		if etcdCfg.AutoCompact < time.Duration(time.Minute*60) {
-			plugin.Log.Warnf("Auto compact option for ETCD is set to less than 60 minutes!")
-		}
-		plugin.startPeriodicAutoCompact(etcdCfg.AutoCompact)
-	}
-	plugin.protoWrapper = kvproto.NewProtoWrapperWithSerializer(plugin.connection, &keyval.SerializerJSON{})
-
-	// Register for providing status reports (polling mode).
-	if plugin.StatusCheck != nil {
-		plugin.StatusCheck.Register(core.PluginName(plugin.PluginName), func() (statuscheck.PluginState, error) {
-			_, _, _, err := plugin.connection.GetValue(healthCheckProbeKey)
-			if err == nil {
-				if plugin.reconnectResync && plugin.lastConnErr != nil {
-					plugin.Log.Info("Starting resync after ETCD reconnect")
-					if plugin.Resync != nil {
-						plugin.Resync.DoResync()
-						plugin.lastConnErr = nil
-					} else {
-						plugin.Log.Warn("Expected resync after ETCD reconnect could not start beacuse of missing Resync plugin")
-					}
+	if err != nil && etcdCfg.DelayedStart {
+		plugin.Log.Infof("ETCD server %s not reachable in init phase. Agent will continue to try to connect", etcdCfg.Endpoints)
+		// Even if the connection cannot be established during init, keep trying
+		go func(etcdCfg *Config) {
+			for {
+				time.Sleep(reconnectInterval * time.Second)
+				plugin.Log.Infof("Connecting to ETCD %v ...", etcdCfg.Endpoints)
+				plugin.connection, err = NewEtcdConnectionWithBytes(*etcdClientCfg, plugin.Log)
+				if err != nil {
+					// Do not mark it as an error
+					plugin.Log.Debugf("Repeated attempt to connect to the ETCD %s failed. Err: %s",
+						etcdCfg.Endpoints, err)
+				} else {
+					plugin.Log.Infof("ETCD server %s connected", etcdCfg.Endpoints)
+					// Configure connection and set as connected
+					plugin.configureConnection(etcdCfg)
+					plugin.connected = true
+					// Notify the dbsync that the database is available and can be registered
+					plugin.initNotifChan <- struct{}{}
+					// Wait until registration is done
+					<-plugin.initNotifChan
```

Comment:
There is a race condition in waiting for registration to be done. It would be better to use callback for the commands that needs to be executed on connect.